### ISSUE TITLE
Getting "ValueError: Pointer argument (at 0) cannot be accessed from Triton (cpu tensor?)" when trying to run Qwen/Qwen2-VL-72B-Instruct-AWQ

### ISSUE THREAD

**Issue by Lissanro** [matatonic/openedai-vision#20]
Once installation completed, I run it like this (I have four 3090 24GB GPUs):

```
.venv/bin/python vision.py --model Qwen/Qwen2-VL-72B-Instruct-AWQ -A flash_attention_2 --device-map auto
```

But then when I try inference:

```
.venv/bin/python chat_with_image.py -1 https://images.freeimages.com/images/large-previews/cd7/gingko-biloba-1058537.jpg "Describe the image."
```

It fails with this error (I also tried without flash_attention_2, but it did not help):

```
ValueError: Pointer argument (at 0) cannot be accessed from Triton (cpu tensor?)
```

Full error log: https://pastebin.com/ZeSPwNAk

**Comment by matatonic** [comment]
Have you tried `--device-map sequential`? I find that works best for me with vision models.

**Comment by Lissanro** [comment]
Yes, I tried it with `--device-map sequential` too, then it fills memory on two GPUs (instead of spreading VRAM usage across all four GPUs), but the error remains the same when I try inference.

In case it matters, I am using Ubuntu 24.04 with Nvidia driver 550.90.07, and this is how I installed openedai-vision:

```
git clone https://github.com/matatonic/openedai-vision.git
cd openedai-vision
wget https://dragon.studio/2024/09/openedai-vision-issue-19.patch
patch -p1 < openedai-vision-issue-19.patch
python -m venv .venv
.venv/bin/pip install -U torch numpy --no-build-isolation
.venv/bin/pip install -U git+https://github.com/AutoGPTQ/AutoGPTQ.git --no-build-isolation
.venv/bin/pip install -U -r requirements.txt --no-build-isolation
.venv/bin/pip install -U git+https://github.com/huggingface/transformers@21fac7abba2a37fae86106f87fcf9974fd1e3830 --no-build-isolation
.venv/bin/pip install -U git+https://github.com/casper-hansen/AutoAWQ.git --no-build-isolation
```

The reason why I am installing specific transformers version is because at the time of writing, there is a bug: https://github.com/huggingface/transformers/issues/33401 .

I hit other issues along the way (for reference: https://github.com/AutoGPTQ/AutoGPTQ/issues/339, https://github.com/AutoGPTQ/AutoGPTQ/issues/500 and https://github.com/matatonic/openedai-vision/issues/19 ) - this is why I disable build isolation and install torch and numpy first, and apply a patch to openedai-vision.

If I did something wrong in my setup, please let me know.

**Comment by matatonic** [comment]
That's a lot of very specific changes... I tested the AWQ with device map auto with the docker image I built and it worked there and passed tests. Have you tried the pre-built docker image? 

there are too many variables with pip, python, driver, etc for me to test with anything other than the docker image. I don't even test it on my own system anymore, just docker.

**Comment by matatonic** [comment]
Frankly, it's a serious challenge to get all the versions and environment working together to support all the different models at the same time for a single docker image. I don't have the resources to test such a specific, customized and complex variant of my build. The error you mentioned with transformers seems to be cosmetic btw. The fix for #19 will be out shortly, final testing is running now and should be done in a couple hours if all goes well.

**Comment by Lissanro** [comment]
I mostly just installed some required packages from git, and applied the fix suggested in https://github.com/matatonic/openedai-vision/issues/19 so I do not think I did much changes. I tried first running commands as suggested in the Readme but all of them resulted in errors, both for manual installation (since it wasn't taking into account all the issues I have encountered) and for the docker. They way I did it was the only way I found to get to at least load the model.

I agree that it may be hard to take into account all possible environments, but it may be worth considering testing them from time to time in the latest LTS version of Ubuntu, as one of the most popular distros. I would not mind helping with that and work on minimizing necessary commands to run for the manual installation, if I can get it up and running first, and start using regularly. Because if I cannot run it, I don't really know yet what is necessary to make it work.

I tried the docker like this: in vision.env I uncommented Qwen2-VL-72B, and ran "docker compose up", but got an error:

```
> docker compose up
[+] Running 19/19
 ✔ openedai-vision Pulled                                                                                                                                                                                                2368.8s 
   ✔ a2318d6c47ec Pull complete                                                                                                                                                                                            49.8s 
   ✔ 40d734479f14 Pull complete                                                                                                                                                                                            49.9s 
   ✔ 0b61b7b259eb Pull complete                                                                                                                                                                                            50.4s 
   ✔ 081a3493c0e7 Pull complete                                                                                                                                                                                            50.5s 
   ✔ 4d14460e4ffb Pull complete                                                                                                                                                                                           119.5s 
   ✔ 8b791ccadd97 Pull complete                                                                                                                                                                                           119.8s 
   ✔ 556302bc119b Pull complete                                                                                                                                                                                           119.8s 
   ✔ a928f71eefed Pull complete                                                                                                                                                                                           158.5s 
   ✔ 10727726de07 Pull complete                                                                                                                                                                                           158.6s 
   ✔ 7dcf93a4fa6f Pull complete                                                                                                                                                                                           158.6s 
   ✔ 4e042b0f2dd1 Pull complete                                                                                                                                                                                           158.6s 
   ✔ 593c3020a5ff Pull complete                                                                                                                                                                                          2365.3s 
   ✔ 4f4fb700ef54 Pull complete                                                                                                                                                                                          2365.3s 
   ✔ 2f6407744529 Pull complete                                                                                                                                                                                          2365.3s 
   ✔ 00c1030ee0c9 Pull complete                                                                                                                                                                                          2365.5s 
   ✔ 006461945655 Pull complete                                                                                                                                                                                          2365.5s 
   ✔ f196f9271fa2 Pull complete                                                                                                                                                                                          2365.5s 
   ✔ 1e5f4ea436de Pull complete                                                                                                                                                                                          2365.5s 
[+] Running 2/2
 ✔ Network openedai-vision_default  Created                                                                                                                                                                                 0.2s 
 ✔ Container openedai-vision        Created                                                                                                                                                                                 1.9s 
Attaching to openedai-vision
Gracefully stopping... (press Ctrl+C again to force)
Error response from daemon: could not select device driver "nvidia" with capabilities: [[gpu]]
```

**Comment by matatonic** [comment]
> Error response from daemon: could not select device driver "nvidia" with capabilities: [[gpu]]

Did you see this note about setting the docker container runtime to be nvidia by default (see: https://github.com/matatonic/openedai-vision?tab=readme-ov-file#known-problems--workarounds)

**Comment by matatonic** [comment]
In the docker-compose.yml you can also set:
```
 runtime: nvidia
```

instead.

**Comment by matatonic** [comment]
Also, I'm certainly not against assistance fixing up the manual build instructions, but I don't want to deviate much from the docker build or requirements, there is already a fork for the older 'alt' build.

**Comment by matatonic** [comment]
FYI, the build is a bust, awq seems busted so I wont be updating yet. I'll push my changes thus far into a dev branch for now while I test the rest.

**Comment by Lissanro** [comment]
Yes, I tried the workaround in the Readme and it had no effect:

```
> cat /etc/docker/daemon.json
{
    "runtimes": {
        "nvidia": {
            "path": "nvidia-container-runtime",
            "runtimeArgs": []
        }
    },
    "default-runtime": "nvidia"
}
```

```
> docker compose up          
Attaching to openedai-vision
Gracefully stopping... (press Ctrl+C again to force)
Error response from daemon: could not select device driver "nvidia" with capabilities: [[gpu]]
```

I also tried edit docker-compose.yml as you have suggested, but then it failed with slightly different error:

```
> cat docker-compose.yml
services:
  openedai-vision:
    build:
      args:
        - VERSION=latest
      dockerfile: Dockerfile
    container_name: openedai-vision
    image: ghcr.io/matatonic/openedai-vision
    env_file: vision.env # your settings go here
    volumes:
      - ./hf_home:/app/hf_home  # for Hugginface model cache
      - ./model_conf_tests.json:/app/model_conf_tests.json
    ports:
      - 5006:5006
    runtime: nvidia
    deploy:
      resources:
        reservations:
          devices:
            - driver: nvidia
              #device_ids: ['0', '1'] # Select a gpu, or
              count: all
              capabilities: [gpu]
```

```
> docker compose up     
[+] Running 0/0
 ⠋ Container openedai-vision  Recreate                                                                                                                                                                                      0.0s 
Error response from daemon: unknown or invalid runtime name: nvidia
```

The readme also mentioned installing "nvidia-container-runtime", but no such package exist in the modern Ubuntu repository:

```
> sudo apt install nvidia-container-runtime
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
E: Unable to locate package nvidia-container-runtime
```

When googling for it, I found the official repository https://github.com/NVIDIA/nvidia-container-runtime and it says it is deprecated and archived in favor of [NVIDIA Container Toolkit](https://github.com/NVIDIA/nvidia-container-toolkit).

I found these instructions to install nvidia-container-toolkit: https://github.com/NVIDIA/nvidia-container-toolkit/issues/482#issuecomment-2117787042 and I was able to successfully install it:

```
> sudo apt-get install -y nvidia-container-toolkit
...
Get:1 https://nvidia.github.io/libnvidia-container/stable/deb/amd64  libnvidia-container1 1.16.1-1 [925 kB]
Get:2 https://nvidia.github.io/libnvidia-container/stable/deb/amd64  libnvidia-container-tools 1.16.1-1 [19.8 kB]
Get:3 https://nvidia.github.io/libnvidia-container/stable/deb/amd64  nvidia-container-toolkit-base 1.16.1-1 [3,511 kB]
Get:4 https://nvidia.github.io/libnvidia-container/stable/deb/amd64  nvidia-container-toolkit 1.16.1-1 [1,128 kB]
...
Setting up nvidia-container-toolkit-base (1.16.1-1) ...
Setting up libnvidia-container1:amd64 (1.16.1-1) ...
Setting up libnvidia-container-tools (1.16.1-1) ...
Setting up nvidia-container-toolkit (1.16.1-1) ...
Processing triggers for libc-bin (2.39-0ubuntu8.3) ...
```

Then I tried again, and the docker still fails:

```
> docker compose up                        
[+] Running 0/0
 ⠋ Container openedai-vision  Recreate                                                                                                                                                                                      0.0s 
Error response from daemon: unknown or invalid runtime name: nvidia
```

Please let me know if there is anything else you would like me to try.

**Comment by matatonic** [comment]
That's looking right, did you restart docker after you changed the config and installed the container toolkit? (you should also be able to install nvidia-container-runtime now)

**Comment by matatonic** [comment]
I also install nvidia-docker2, but this may be obsolete and not required anymore.

**Comment by Lissanro** [comment]
"sudo service docker restart" was the command I missed. Neither nvidia-container-runtime nor nvidia-docker2 are needed I think, since they are deprecated packages, so I did not install them. I think only these steps were necessary:

- Install `nvidia-container-toolkit` using instructions from https://github.com/NVIDIA/nvidia-container-toolkit/issues/482#issuecomment-2117787042
- Uncomment `runtime: nvidia` in `docker-compose.yml`
- Run `sudo service docker restart`
- Run `docker compose up`

```
> docker compose up
[+] Running 1/1
 ✔ Container openedai-vision  Recreated                                                                                                                                                                                     0.1s 
Attaching to openedai-vision
openedai-vision  | 2024-09-22 16:27:09.891 | INFO     | __main__:<module>:146 - Loading VisionQnA[qwen2-vl] with Qwen/Qwen2-VL-72B-Instruct-AWQ
openedai-vision  | Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'mrope_section'}
openedai-vision  | We suggest you to set `torch_dtype=torch.float16` for better efficiency with AWQ.
Downloading shards:   0%|          | 0/11 [01:33<?, ?it/s]
openedai-vision  | Traceback (most recent call last):
openedai-vision  |   File "/usr/local/lib/python3.11/site-packages/huggingface_hub/file_download.py", line 515, in http_get
openedai-vision  |     hf_transfer.download(
openedai-vision  | Exception: Failed too many failures in parallel (3): Exception: Error while downloading: error sending request for url (https://cdn-lfs-us-1.huggingface.co/repos/fd/...
```

It fails to download the model, not sure what is needed to enable internet connection with docker, but in this case I would not want to download it anyway, because I already have it at ~/.cache/huggingface/hub/models--Qwen--Qwen2-VL-72B-Instruct-AWQ (it would take very long time to download it again with my internet connection) - but it seems the docker container cannot see it. I am at the moment looking for a way to fix this or copy files inside of it somehow.

**Comment by matatonic** [comment]
fails to download, that's new. Did you adjust the `HF_HOME` variable in images.env?

**Comment by matatonic** [comment]
you will need to map the local folder into the container as a volume, by default this is `./hf_home`

**Comment by Lissanro** [comment]
By the way, "Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'mrope_section'}" error means transformers version in the docker container maybe incorrect and hitting this issue https://github.com/huggingface/transformers/issues/33401 - so perhaps I will need to run `pip install git+https://github.com/huggingface/transformers@21fac7abba2a37fae86106f87fcf9974fd1e3830` inside the docker container to solve this, just like was the case when doing the manual install.

**Comment by matatonic** [comment]
that error doesn't stop it from working in my experience.
just fyi, I have so far been unable to build a new working version, and have tried a few variations of transformers git versions, yours included.

**Comment by matatonic** [comment]
As a test, you could:
```
mv  ~/.cache/huggingface/hub hf_home/hub
```

**Comment by Lissanro** [comment]
I do not have images.env, maybe you meant vision.env? Here is content of vision.env:

```
HF_HOME=hf_home
HF_HUB_ENABLE_HF_TRANSFER=1
CLI_COMMAND="python vision.py -m Qwen/Qwen2-VL-72B-Instruct-AWQ -A flash_attention_2"  # test pass✅, time: 35.2s, mem: 44.5GB, 13/13 tests passed.
```

But I think I get it now, I need to map hf_home to my real huggingface cache, I run these commands:

```
sudo rm -rf hf_home && ln -s ~/.cache/huggingface/ $(pwd)/hf_home
```

Alternatively, I could have edited the vision.env file to specify path to ~/.cache/huggingface.

After that, I tried again:

```
> docker compose up
[+] Running 1/0
 ✔ Container openedai-vision  Created                                                                                                                                                                                       0.0s 
Attaching to openedai-vision
openedai-vision  | 2024-09-22 16:51:33.049 | INFO     | __main__:<module>:146 - Loading VisionQnA[qwen2-vl] with Qwen/Qwen2-VL-72B-Instruct-AWQ
openedai-vision  | Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'mrope_section'}
openedai-vision  | We suggest you to set `torch_dtype=torch.float16` for better efficiency with AWQ.
openedai-vision  | `Qwen2VLRotaryEmbedding` can now be fully parameterized by passing the model config through the `config` argument. All other arguments will be removed in v4.46
Loading checkpoint shards: 100%|██████████| 11/11 [01:22<00:00,  7.47s/it]
openedai-vision  | 2024-09-22 16:53:03.131 | INFO     | vision_qna:loaded_banner:90 - Loaded Qwen/Qwen2-VL-72B-Instruct-AWQ [ device: cuda:0, dtype: torch.bfloat16, template: None ]
openedai-vision  | INFO:     Started server process [7]
openedai-vision  | INFO:     Waiting for application startup.
openedai-vision  | INFO:     Application startup complete.
openedai-vision  | INFO:     Uvicorn running on http://0.0.0.0:5006 (Press CTRL+C to quit)
openedai-vision  | INFO:     172.19.0.1:54316 - "POST /v1/chat/completions HTTP/1.1" 200 OK
openedai-vision  | /usr/local/lib/python3.11/site-packages/transformers/generation/configuration_utils.py:604: UserWarning: `do_sample` is set to `False`. However, `top_p` is set to `0.001` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `top_p`.
openedai-vision  |   warnings.warn(
```

While running in other terminal inference test:

```
> .venv/bin/python chat_with_image.py -1 https://images.freeimages.com/images/large-previews/cd7/gingko-biloba-1058537.jpg "Describe the image."
The image shows a single leaf, which appears to be from a ginkgo tree. The leaf has a distinctive fan-like shape with a broad, flat surface that tapers to a point at the base where it attaches to the stem. The edges of the leaf are slightly wavy, and the surface is smooth with visible veins radiating outward from the central stem. The color of the leaf is a vibrant green, indicating that it is likely healthy and alive.
```

It worked, even without solving the `Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'mrope_section'}` error (I tried `docker exec -it openedai-vision pip install git+https://github.com/huggingface/transformers@21fac7abba2a37fae86106f87fcf9974fd1e3830` but it did not make it go away, but at least it seems to be working anyway). The memory usage spread out across all 4 GPUs as excpected, using around 10-14GB on each.

Thank you for your help so far. I hope steps for getting the docker container up and running will be useful to update the Readme.

The original issue of this bug report with the manual installation not working is not solved yet, but now that I got it up and running with the docker container, and have a working configuration, I will try to determine correct steps for manual installation within next few days, and report back (even if I can't find working manual steps, I will at least provide what else I tried, for reference).

EDIT: Since I did not build the docker container from scratch, I forgot to mention additional step - I had to copy patched `./backend/qwen2-vl.py` to replace incorrect `/app/backend/qwen2-vl.py` inside the docker container (to solve the issue #19).

**Comment by matatonic** [comment]
fyi, some recent model config changes have broken qwen-vl-7b-awq for me:

https://huggingface.co/Qwen/Qwen2-VL-7B-Instruct-AWQ/commit/1dca9854ff028edf7df85575c4e86546f7faa9df

before this revision is ok.

**Comment by matatonic** [comment]
Going to close this for now. If you get time to re-open a bug related to manual install instructions not working please do so. Thanks again!

**Comment by Lissanro** [comment]
Not sure why you closed this, nothing has been fixed (except issue #19 which does not need a workaround anymore, but other steps still need updating):

- Manual instructions are still broken in the Readme. Steps I shared seem to have issues with getting tensor parallelism working, but at least they get it running and likely will work for models that require only a single GPU. I did not find yet steps to successfully run it manually with tensor parallelism.

- Steps getting the docker container working are also still broken in Readme, you did not update them even though I shared up to date steps: https://github.com/matatonic/openedai-vision/issues/20#issuecomment-2366865831 (only missing last step was either using local cache as described here: https://github.com/matatonic/openedai-vision/issues/20#issuecomment-2366881798 or finding a way to enable it to download it from within the container).

Only step that can be excluded is what was related to workaround for the #19 issue, which is not needed anymore.

I cannot re-open the bug (github offers no such option for me), so there is nothing I can do. Closing unfixed bug will make it harder for people to find a solution, so if you don't have time to fix Readme right now, I suggest just keeping this bug open until actually fixed and steps are verified to work on modern distros such as Ubuntu 24.04 (I shared steps that worked for me to one extend or another; if I find better steps, I will share them).

**Comment by matatonic** [comment]
I closed this ticket because it seemed you were able to get the model working. I'd rather have issue specific tickets, with solutions rather than a large collection of unrelated issues. 

The `runtime: nvidia` is a container runtime setup issue and not related to this project specifically. Setting up docker with nvidia is beyond the scope of this project and I updated the link to the setup guides for the nvidia container toolkit in the docs. If you follow the instructions there, nvidia will be the detault container runtime and setting runtime: nvidia is not required for linux and not possible for windows. If you don't want to set the default this way, as nvidia recommends, that's fine but it's not for everyone.

You say that the manual install is broken, but perhaps I missed it, I don't see which step is broken. Can you tell me which failure you get and which step is wrong?

The release has been updated and git transformers is no longer required for it to work. I suggest you try the instructions (or new docker) again and update the ticket with any issues you have.

You say tensor parallelism, I'm not aware that this can work in this setup - do you have any information about where you find this working? As far as I know transformers is only able to manage splitting the model across multiple gpu using accelerate.

**Comment by matatonic** [comment]
With no further update, I'm closing this issue.

**Comment by Lissanro** [comment]
Apologies for not updating sooner, but I just do not have new information to add at the moment. The bug in the title was still there last time I checked and manual installation instructions were still broken in the Readme. Please let me know if there is anything specific you would like me to try, or if you want me retest again, or if you need additional information that could help. I already provided very detailed description for my steps, and at this point I am not sure how to debug this further. I tried but could not discover any new information yet beyond what I already provided in this bug report. I will try to summarize everything below.

This bug is about the issue in its title using manual installation method. That said, you suggested to also try docker, so I investigated this as well, but I think I already discovered minimal steps to get docker working, but to keep things more organized, I can create a new issue about it. What trying docker way allowed to discover though, that the issue that this bug report is about should be solvable also in theory.

So, the original issue comes down to two problems:

- Current manual installation instructions in Readme are not working

- Steps that I have discovered to get it to run using the manual installation method result in the error in the title of this bug report (but since the error does not happen whet installed with docker, I assume I missed something)

As of the manual installation, these are the steps that worked for me (I already mentioned before what bugs I hit specifically that forced me to do these additional steps, and provided links to them):

git clone https://github.com/matatonic/openedai-vision.git
cd openedai-vision
python -m venv .venv
.venv/bin/pip install -U torch numpy --no-build-isolation
.venv/bin/pip install -U git+https://github.com/AutoGPTQ/AutoGPTQ.git --no-build-isolation
.venv/bin/pip install -U -r requirements.txt --no-build-isolation
.venv/bin/pip install -U git+https://github.com/huggingface/transformers@21fac7abba2a37fae86106f87fcf9974fd1e3830 --no-build-isolation
.venv/bin/pip install -U git+https://github.com/casper-hansen/AutoAWQ.git --no-build-isolation

I find installation without a docker more practical to experiment with, it would be great to make it work with multiple GPUs. I do not know if tensor parallelism is the right term, maybe the backend library openedai-vision is using calls usage of multiple GPUs differently. But the issue is, as the title of this bug reports indicates, it fails to run on multiple GPUs, but succeeds in the docker.

Please let me know if there is anything else you would like me to try, or if there is any additional information is needed from running a specific command(s). I already described what issues I encountered during the manual installation method in this message following steps in the Readme: https://github.com/matatonic/openedai-vision/issues/20#issuecomment-2366773416 .

**Comment by matatonic** [comment]
I have a couple theories why your manual install didn't work, but I'd like to reproduce the problem first. Your solution may work for you, which is great, but doesn't really help narrow down and identify the problem, it's just too many changes.
Some things I realized so far after review:
- I don't upgrade pip before the install, this is probably required but maybe not your issue
- flash_attn only links to pre-built modules for cuda 12.3 and python 3.10,3.11. building a new module requires a lot more to be installed.

If you're willing to assist, I'd like to have you start over and try this:
```shell
git clone https://github.com/matatonic/openedai-vision
cd openedai-vision
python -m venv .venv
source .venv/bin/activate
pip install -U -r requirements.txt "transformers>=4.45.1"
python debug_info.py
```

It should give you some output like this (from my ubuntu 22 system):
```
---- cuda
CUDA version: 12.1
torch.cuda.is_available(): True
CUDA device[0](8, 9): NVIDIA GeForce RTX 4090
CUDA device[1](8, 0): NVIDIA A100 80GB PCIe
---- python
3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0]
---- pip
accelerate==1.0.0
auto_gptq==0.7.1
autoawq==0.2.5
bitsandbytes==0.44.1
decord==0.6.0
deepspeed==0.15.2
einops==0.8.0
einops-exts==0.0.4
fastapi==0.115.0
flash_attn==2.6.3
hf_transfer==0.1.8
httpx==0.27.2
logger==1.4
loguru==0.7.2
markdown2==2.5.1
matplotlib==3.9.2
mistral_common==1.4.4
mistral_inference==1.4.0
numpy==2.1.2
open_clip_torch==2.26.1
openai==1.51.2
optimum==1.23.1
peft==0.13.2
protobuf==5.28.2
pydantic==2.9.2
python-datauri==2.2.0
qwen-vl-utils==0.0.8
requests==2.32.3
sentencepiece==0.2.0
shortuuid==1.0.13
sse_starlette==2.1.3
tiktoken==0.7.0
timm==1.0.9
tokenizers==0.20.1
torch==2.4.1
torchvision==0.19.1
transformers==4.45.2
transformers_stream_generator==0.0.5
uvicorn==0.31.1
verovio==4.3.1
wandb==0.18.3
xformers==0.0.28.post1
```

Then try to run the server.
```
# optional exports, depending on your setup:
export HF_HOME=hf_home
python vision.py -m Qwen/Qwen2-VL-72B-Instruct-AWQ -A flash_attention_2"
```
For me this loaded across 2 devices.

Then try inference from another terminal:
```
python chat_with_image.py -1 https://images.freeimages.com/images/large-previews/cd7/gingko-biloba-1058537.jpg "Describe the image."
```

Which gave me output at about  5 T/s, which isn't great, but it worked. Please let me know how this goes for you.

**Comment by Lissanro** [comment]
These steps do not work. After cloning and initializing venv, this is what I get (tried both without and with upgrading pip):

```
> source .venv/bin/activate
(.venv) lissanro@pc /t/openedai-vision> pip install -U -r requirements.txt "transformers>=4.45.1"
Ignoring flash-attn: markers 'platform_system == "Linux" and platform_machine == "x86_64" and python_version == "3.11"' don't match your environment
Ignoring flash-attn: markers 'platform_system == "Linux" and platform_machine == "x86_64" and python_version == "3.10"' don't match your environment
Ignoring flash-attn: markers 'platform_system == "Windows" and python_version == "3.11"' don't match your environment
Ignoring flash-attn: markers 'platform_system == "Windows" and python_version == "3.10"' don't match your environment
Collecting git+https://github.com/LLaVA-VL/LLaVA-NeXT.git (from -r requirements.txt (line 55))
  Cloning https://github.com/LLaVA-VL/LLaVA-NeXT.git to /tmp/pip-req-build-qm510m1s
  Running command git clone --filter=blob:none --quiet https://github.com/LLaVA-VL/LLaVA-NeXT.git /tmp/pip-req-build-qm510m1s
  Resolved https://github.com/LLaVA-VL/LLaVA-NeXT.git to commit 03b70e2d69936d069f54694ce3eb57538ef598f3
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting transformers>=4.45.1
  Downloading transformers-4.45.2-py3-none-any.whl.metadata (44 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.4/44.4 kB 133.8 kB/s eta 0:00:00
Collecting accelerate (from -r requirements.txt (line 1))
  Downloading accelerate-1.0.0-py3-none-any.whl.metadata (19 kB)
Collecting auto_gptq (from -r requirements.txt (line 2))
  Using cached auto_gptq-0.7.1.tar.gz (126 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [1 lines of output]
      Building cuda extension requires PyTorch (>=1.13.0) being installed, please install PyTorch first: No module named 'torch'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

Installing torch first is a necessary step. Also `pip install -U -r requirements.txt "transformers>=4.45.1"` needs `--no-build-isolation` flag, or it will fail again with the same error as above. It will complain about missing numpy next:

```
> pip install -U -r requirements.txt "transformers>=4.45.1" --no-build-isolation
Ignoring flash-attn: markers 'platform_system == "Linux" and platform_machine == "x86_64" and python_version == "3.11"' don't match your environment
Ignoring flash-attn: markers 'platform_system == "Linux" and platform_machine == "x86_64" and python_version == "3.10"' don't match your environment
Ignoring flash-attn: markers 'platform_system == "Windows" and python_version == "3.11"' don't match your environment
Ignoring flash-attn: markers 'platform_system == "Windows" and python_version == "3.10"' don't match your environment
Collecting git+https://github.com/LLaVA-VL/LLaVA-NeXT.git (from -r requirements.txt (line 55))
  Cloning https://github.com/LLaVA-VL/LLaVA-NeXT.git to /tmp/pip-req-build-rfkosgdr
  Running command git clone --filter=blob:none --quiet https://github.com/LLaVA-VL/LLaVA-NeXT.git /tmp/pip-req-build-rfkosgdr
  Resolved https://github.com/LLaVA-VL/LLaVA-NeXT.git to commit 03b70e2d69936d069f54694ce3eb57538ef598f3
  Preparing metadata (pyproject.toml) ... done
Collecting transformers>=4.45.1
  Using cached transformers-4.45.2-py3-none-any.whl.metadata (44 kB)
Collecting accelerate (from -r requirements.txt (line 1))
  Using cached accelerate-1.0.0-py3-none-any.whl.metadata (19 kB)
Collecting auto_gptq (from -r requirements.txt (line 2))
  Using cached auto_gptq-0.7.1.tar.gz (126 kB)
  Preparing metadata (pyproject.toml) ... done
Discarding https://files.pythonhosted.org/packages/90/e5/b22697903982284fe284568fb2663a2196694a8eee637f5cf4ccfe435a38/auto_gptq-0.7.1.tar.gz (from https://pypi.org/simple/auto-gptq/) (requires-python:>=3.8.0): Requested auto_gptq from https://files.pythonhosted.org/packages/90/e5/b22697903982284fe284568fb2663a2196694a8eee637f5cf4ccfe435a38/auto_gptq-0.7.1.tar.gz (from -r requirements.txt (line 2)) has inconsistent version: expected '0.7.1', but metadata has '0.7.1+cu121'
  Using cached auto_gptq-0.7.0.tar.gz (124 kB)
  Preparing metadata (pyproject.toml) ... done
Discarding https://files.pythonhosted.org/packages/34/71/c3e73cf17681f6ff4754ef8f4cb8b67af3def230fc8711eac1250bbd78d5/auto_gptq-0.7.0.tar.gz (from https://pypi.org/simple/auto-gptq/) (requires-python:>=3.8.0): Requested auto_gptq from https://files.pythonhosted.org/packages/34/71/c3e73cf17681f6ff4754ef8f4cb8b67af3def230fc8711eac1250bbd78d5/auto_gptq-0.7.0.tar.gz (from -r requirements.txt (line 2)) has inconsistent version: expected '0.7.0', but metadata has '0.7.0+cu121'
  Using cached auto_gptq-0.6.0.tar.gz (120 kB)
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [32 lines of output]
      /tmp/openedai-vision/.venv/lib/python3.12/site-packages/torch/_subclasses/functional_tensor.py:258: UserWarning: Failed to initialize NumPy: No module named 'numpy' (Triggered internally at ../torch/csrc/utils/tensor_numpy.cpp:84.)
        cpu = _conversion_method_template(device=torch.device("cpu"))
      python: can't open file '/tmp/pip-install-j4dlpd82/auto-gptq_18a3b784797d486999279c0122b4d466/./autogptq_extension/qigen/generate.py': [Errno 2] No such file or directory
      Generating qigen kernels...
      Traceback (most recent call last):
        File "<string>", line 109, in <module>
        File "/usr/lib/python3.12/subprocess.py", line 466, in check_output
          return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/usr/lib/python3.12/subprocess.py", line 571, in run
          raise CalledProcessError(retcode, process.args,
      subprocess.CalledProcessError: Command '['python', './autogptq_extension/qigen/generate.py', '--module', '--search', '--p', '16']' returned non-zero exit status 2.
      
      During handling of the above exception, another exception occurred:
      
      Traceback (most recent call last):
        File "/tmp/openedai-vision/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/tmp/openedai-vision/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/openedai-vision/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 149, in prepare_metadata_for_build_wheel
          return hook(metadata_directory, config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/openedai-vision/.venv/lib/python3.12/site-packages/setuptools/build_meta.py", line 373, in prepare_metadata_for_build_wheel
          self.run_setup()
        File "/tmp/openedai-vision/.venv/lib/python3.12/site-packages/setuptools/build_meta.py", line 503, in run_setup
          super().run_setup(setup_script=setup_script)
        File "/tmp/openedai-vision/.venv/lib/python3.12/site-packages/setuptools/build_meta.py", line 318, in run_setup
          exec(code, locals())
        File "<string>", line 111, in <module>
      Exception: Generating QiGen kernels failed with the error shown above.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

After installing numpy, it will complain about AutoGPTQ:

```
(.venv) lissanro@pc /t/openedai-vision> pip install -U -r requirements.txt "transformers>=4.45.1" --no-build-isolation
Ignoring flash-attn: markers 'platform_system == "Linux" and platform_machine == "x86_64" and python_version == "3.11"' don't match your environment
Ignoring flash-attn: markers 'platform_system == "Linux" and platform_machine == "x86_64" and python_version == "3.10"' don't match your environment
Ignoring flash-attn: markers 'platform_system == "Windows" and python_version == "3.11"' don't match your environment
Ignoring flash-attn: markers 'platform_system == "Windows" and python_version == "3.10"' don't match your environment
Collecting git+https://github.com/LLaVA-VL/LLaVA-NeXT.git (from -r requirements.txt (line 55))
  Cloning https://github.com/LLaVA-VL/LLaVA-NeXT.git to /tmp/pip-req-build-b261wokk
  Running command git clone --filter=blob:none --quiet https://github.com/LLaVA-VL/LLaVA-NeXT.git /tmp/pip-req-build-b261wokk
  Resolved https://github.com/LLaVA-VL/LLaVA-NeXT.git to commit 03b70e2d69936d069f54694ce3eb57538ef598f3
  Preparing metadata (pyproject.toml) ... done
Collecting transformers>=4.45.1
  Using cached transformers-4.45.2-py3-none-any.whl.metadata (44 kB)
Collecting accelerate (from -r requirements.txt (line 1))
  Using cached accelerate-1.0.0-py3-none-any.whl.metadata (19 kB)
Collecting auto_gptq (from -r requirements.txt (line 2))
  Using cached auto_gptq-0.7.1.tar.gz (126 kB)
  Preparing metadata (pyproject.toml) ... done
Discarding https://files.pythonhosted.org/packages/90/e5/b22697903982284fe284568fb2663a2196694a8eee637f5cf4ccfe435a38/auto_gptq-0.7.1.tar.gz (from https://pypi.org/simple/auto-gptq/) (requires-python:>=3.8.0): Requested auto_gptq from https://files.pythonhosted.org/packages/90/e5/b22697903982284fe284568fb2663a2196694a8eee637f5cf4ccfe435a38/auto_gptq-0.7.1.tar.gz (from -r requirements.txt (line 2)) has inconsistent version: expected '0.7.1', but metadata has '0.7.1+cu121'
  Using cached auto_gptq-0.7.0.tar.gz (124 kB)
  Preparing metadata (pyproject.toml) ... done
Discarding https://files.pythonhosted.org/packages/34/71/c3e73cf17681f6ff4754ef8f4cb8b67af3def230fc8711eac1250bbd78d5/auto_gptq-0.7.0.tar.gz (from https://pypi.org/simple/auto-gptq/) (requires-python:>=3.8.0): Requested auto_gptq from https://files.pythonhosted.org/packages/34/71/c3e73cf17681f6ff4754ef8f4cb8b67af3def230fc8711eac1250bbd78d5/auto_gptq-0.7.0.tar.gz (from -r requirements.txt (line 2)) has inconsistent version: expected '0.7.0', but metadata has '0.7.0+cu121'
  Using cached auto_gptq-0.6.0.tar.gz (120 kB)
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [30 lines of output]
      python: can't open file '/tmp/pip-install-lxucsz0b/auto-gptq_39071b1e688b4ee396e9bdb2c5901aa3/./autogptq_extension/qigen/generate.py': [Errno 2] No such file or directory
      Generating qigen kernels...
      Traceback (most recent call last):
        File "<string>", line 109, in <module>
        File "/usr/lib/python3.12/subprocess.py", line 466, in check_output
          return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/usr/lib/python3.12/subprocess.py", line 571, in run
          raise CalledProcessError(retcode, process.args,
      subprocess.CalledProcessError: Command '['python', './autogptq_extension/qigen/generate.py', '--module', '--search', '--p', '16']' returned non-zero exit status 2.
      
      During handling of the above exception, another exception occurred:
      
      Traceback (most recent call last):
        File "/tmp/openedai-vision/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/tmp/openedai-vision/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/openedai-vision/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 149, in prepare_metadata_for_build_wheel
          return hook(metadata_directory, config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/openedai-vision/.venv/lib/python3.12/site-packages/setuptools/build_meta.py", line 373, in prepare_metadata_for_build_wheel
          self.run_setup()
        File "/tmp/openedai-vision/.venv/lib/python3.12/site-packages/setuptools/build_meta.py", line 503, in run_setup
          super().run_setup(setup_script=setup_script)
        File "/tmp/openedai-vision/.venv/lib/python3.12/site-packages/setuptools/build_meta.py", line 318, in run_setup
          exec(code, locals())
        File "<string>", line 111, in <module>
      Exception: Generating QiGen kernels failed with the error shown above.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

So I have to install them first (torch and numpy can be install without the `--no-build-isolation` flag, but both `pip install git+https://github.com/AutoGPTQ/AutoGPTQ.git` and `pip install -U -r requirements.txt "transformers>=4.45.1"` need it to succeed, otherwise they will not see the torch package):

```
pip install -U torch numpy
pip install git+https://github.com/AutoGPTQ/AutoGPTQ.git --no-build-isolation
pip install -U -r requirements.txt "transformers>=4.45.1" --no-build-isolation
```

Only then it becomes possible to proceed with the next steps. Please let me know if you want me to go forward with other steps, or if these first steps need to be done differently before moving on further (I can continue tomorrow).

**Comment by matatonic** [comment]
so this is very strange right, because torch and numpy are in the requirements.txt file. Can you please just stop fixing things after the first error and run the debug_info.py command and send me the results? I need to see the broken system.

**Comment by Lissanro** [comment]
```
(.venv) lissanro@pc /t/openedai-vision> python debug_info.py
/tmp/openedai-vision/debug_info.py:6: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  import pkg_resources
---- cuda
CUDA version: 12.1
torch.cuda.is_available(): True
CUDA device[0](8, 6): NVIDIA GeForce RTX 3090
CUDA device[1](8, 6): NVIDIA GeForce RTX 3090
CUDA device[2](8, 6): NVIDIA GeForce RTX 3090
CUDA device[3](8, 6): NVIDIA GeForce RTX 3090
---- python
3.12.3 (main, Sep 11 2024, 14:17:37) [GCC 13.2.0]
---- pip
accelerate==1.0.0
auto_gptq==0.8.0.dev0+cu121
autoawq: Not found
bitsandbytes: Not found
decord: Not found
deepspeed: Not found
einops: Not found
einops-exts: Not found
fastapi: Not found
flash_attn: Not found
hf_transfer: Not found
httpx: Not found
logger: Not found
loguru: Not found
markdown2: Not found
matplotlib: Not found
mistral_common: Not found
mistral_inference: Not found
numpy==2.1.2
open_clip_torch: Not found
openai: Not found
optimum: Not found
peft==0.13.2
protobuf: Not found
pydantic: Not found
python-datauri: Not found
qwen-vl-utils: Not found
requests==2.32.3
sentencepiece==0.2.0
shortuuid: Not found
sse_starlette: Not found
tiktoken: Not found
timm: Not found
tokenizers==0.20.1
torch==2.4.1
torchvision: Not found
transformers==4.45.2
transformers_stream_generator: Not found
uvicorn: Not found
verovio: Not found
wandb: Not found
xformers: Not found
```

**Comment by Lissanro** [comment]
I am using Ubuntu 24.04 LTS.

**Comment by matatonic** [comment]
Thanks, it's python 3.12 and lack of wheels there, that's why it fails all over.

**Comment by matatonic** [comment]
Until more wheels exist for python 3.12, I probably wont be able to support this yet. I suggest for your venv you use python 3.11 or 3.10  if you still want to run it locally.

**Comment by matatonic** [comment]
I'll keep poking at 3.12 in general, and I am eager to see 3.13 but it's slow going as libs get support.

**Comment by Lissanro** [comment]
OK, thanks, I will try and report back tomorrow.

In the meantime, I was able to create venv with specific Python version like this:

```
virtualenv --python=/usr/bin/python3.11 .venv
```

(`python3.11 -m venv .venv` did not work because it requires `python3.11-venv` and suggests running `sudo apt install python3.11-venv` but trying it results in an error that there is no such installation candidate).

My Internet connection is not very fast and `pip install -U -r requirements.txt "transformers>=4.45.1"` seems to want to download a lot of stuff, so like I mentioned I will continue tomorrow and check if it worked. Hopefully `virtualenv --python=/usr/bin/python3.11 .venv` was the missing step and it will start working.

**Comment by matatonic** [comment]
totally understand, but I can tell you now the docker (which is python 3.11) works, and I know 3.12 doesn't work (autogptq and autoawq don't have python 3.12 wheels yet). Creating a venv for 3.11 or 3.10 from scratch will download probably as much as the docker image, so I don't suggest you bother if you already have that.

**Comment by matatonic** [comment]
I'll update the readme with some 3.12 info in the next release, thanks! I'll create a new ticket for 3.12 support as well, with the current info.

**Comment by matatonic** [comment]
FYI, transformers doesn't require the git patch as of 4.45.2.

**Comment by matatonic** [comment]
updated: https://github.com/matatonic/openedai-vision/commit/a64740fb7187c71ba62413a3049636ea940d87f8
Can you confirm it works?

**Comment by Lissanro** [comment]
Yes, I just tested and it worked. Indeed, the `virtualenv --python=/usr/bin/python3.11 .venv` command was the step that was missing for the manual installation. With it, manual installation and inference across 4 3090 GPUs worked, I tested with this command and model:

```
python vision.py -m Qwen/Qwen2-VL-72B-Instruct-AWQ -A flash_attention_2
```

I think it is safe to close this issue as resolved, since we found the cause of the bug and the fix. Thank you very much for your time and effort in resolving this issue.
